File: tasks/implementations/node/tasks/producer_consumer.js
const fs = require('fs');
const MOD = 4294967291n;
const CAPACITY = 64;

module.exports = function(size, fixturesRoot) {
  const values = fs.readFileSync(`${fixturesRoot}/generated/producer_consumer/${size}.txt`, 'utf8').trim().split(/\s+/).map(Number);
  const buffer = new Array(CAPACITY).fill(0);
  let head = 0, tail = 0, count = 0, produced = 0, consumed = 0;
  let checksum = 0n;
  while (produced < values.length || count > 0) {
    while (produced < values.length && count < CAPACITY) {
      buffer[tail] = values[produced];
      tail = (tail + 1) % CAPACITY;
      produced += 1;
      count += 1;
    }
    if (count > 0) {
      const value = buffer[head];
      head = (head + 1) % CAPACITY;
      count -= 1;
      checksum = (checksum + BigInt(value) * BigInt(consumed + 1)) % MOD;
      consumed += 1;
    }
  }
  console.log(checksum.toString());
};

File: tasks/implementations/rust/tasks/producer_consumer.rs
use std::fs;

pub fn run(size: &str, fixtures_root: &str) {
    const MOD: u64 = 4_294_967_291;
    const CAPACITY: usize = 64;
    let text = fs::read_to_string(format!("{}/generated/producer_consumer/{}.txt", fixtures_root, size)).unwrap();
    let values: Vec<u64> = text.split_whitespace().map(|x| x.parse::<u64>().unwrap()).collect();
    let mut buffer = [0u64; CAPACITY];
    let (mut head, mut tail, mut count, mut produced) = (0usize, 0usize, 0usize, 0usize);
    let mut consumed = 0u64;
    let mut checksum = 0u64;
    while produced < values.len() || count > 0 {
        while produced < values.len() && count < CAPACITY {
            buffer[tail] = values[produced];
            tail = (tail + 1) % CAPACITY;
            produced += 1;
            count += 1;
        }
        if count > 0 {
            let value = buffer[head];
            head = (head + 1) % CAPACITY;
            count -= 1;
            checksum = (checksum + value * (consumed + 1)) % MOD;
            consumed += 1;
        }
    }
    println!("{}", checksum);
}
